Auto-dampening feedback fixes

autoSteve opened the pull request:
* Allow a periodically resetting generation sensor (this likely will not break anything)
* Allow Watt-hour, kilowatt-hour or megawatt-hour generation/site export entities

BJReplay commented:
> I got nothing, and can find nothing

I couldn't see it, either.

BJReplay commented on tests/__init__.py:
```diff
@@ -299,46 +312,108 @@ async def async_setup_aioresponses() -> None:
+            _uom = "kWh"
+    _DAYS = 1
+
+    adjustment = {"kWh": 1.0, "MWh": 1000.0, "Wh": 0.001, "MJ": 1.0, "": 1.0}
```
Dodgy adjustment for dodgy factor 🤣

autoSteve replied:
Tee hee.

BJReplay commented:
Ready to go, I reckon, and do the templates separately.

autoSteve commented:
I reckon this wants more thought, @BJReplay...

My middle-of-day factors that are being introduced are on the back of a single generation interval.

I think this is likely because these single values are from a day where my solar export was _not_ curtailed and it was a reasonably good day. These are incorrect.

Way back I was requiring to have the same number of generation values as "good day" actuals, and it was too aggresive in certain circumstances. Maybe there should be a "majority" of generation points (i.e >3 for 13:00, >2 for 13:30)?

I will code that and see how it runs.

```
Interval 13:00 has peak estimated actual 4.836 and 6 matching intervals: 09-01, 09-05, 09-06, 09-08, 09-11, 09-12
Interval 13:00 max generation: 3.800, [3.8]
Auto-dampen factor for 13:00 is 0.786
Interval 13:30 has peak estimated actual 4.428 and 4 matching intervals: 09-05, 09-06, 09-09, 09-11
Interval 13:30 max generation: 3.600, [3.6]
Auto-dampen factor for 13:30 is 0.813
Interval 14:00 has peak estimated actual 4.017 and 4 matching intervals: 09-05, 09-06, 09-09, 09-11
Interval 14:00 max generation: 2.900, [2.9]
Auto-dampen factor for 14:00 is 0.722
Interval 14:30 has peak estimated actual 3.463 and 4 matching intervals: 09-05, 09-06, 09-09, 09-11
Interval 14:30 max generation: 3.000, [3.0]
Auto-dampen factor for 14:30 is 0.866
```

autoSteve commented:
Yeah. Much better...

<img width="475" height="394" alt="image" src="https://github.com/user-attachments/assets/62c4029d-143b-4e59-8f31-d1de4d3e8e44" />

autoSteve commented:
`pytest` code needs a tweak now.

BJReplay commented:
> Yeah. Much better...

At least you exceed your 10% / 90% POE - me - not so much:

<img width="485" height="398" alt="image" src="https://github.com/user-attachments/assets/1362807a-5322-4f0b-bb4a-f74a317b81e8" />

I'm likely to be below my "don't dump to Zero Hero" threshold of 85% by 6pm.

autoSteve commented:
> `pytest` code needs a tweak now.

`pytest` code is tweaked.

It is also performance optimised in some `__init__` places that are "expensive" to run.

Test runtime is the same as it was, despite building more data. It's slow-ish for auto-dampen, but the code is tested properly with integration tests, and not just unit tests. Entity history actually gets built using synthetic time, and that is expensive to do for entities that "jump" by 0.1 kWh every so often.

I think this code is now solid and 'sendable'.

The example in 'TEMPLATES.md` will not need a release because `README.md` is not manhandled, so that can be merged any time.

BJReplay reviewed (approved):
I am away for work again until Friday from early tomorrow.

LGTM.

Feel free to send it when you are confident, and ping me to promote release discussion to announcement.

gcoan commented:
> Another _suggested tweak_, borrowing from your work.

looks good.   I thought the comparison was of actual to estimated generation, but you've seen the code more than me!  Shows the importance of being explicit and spelling everything out

so people can still not understand but can be pointed to the readme

there are a lot of concepts going on here

cheers

autoSteve commented:
Thanks for the proof read, @Nilogax.

BJReplay commented:
> Thanks for the proof read, @Nilogax.

I've gotta say that I think that the explanation is getting more understandable through the collective efforts of the author and the reviews, so thanks, all.

autoSteve commented:
A late breaking fix.

I accidentally published my API key tonight, so then I changed my API key, thinking, "This'll be a good test". And it was. This should not have resulted in an API key usage tracking reset because nothing else changed, yet it did result in a reset.

This is now fixed, @BJReplay. What's in an indent, eh? Yeah, significance.

This still slips through `pytest` code, but it is such a corner case that my giveadamn is kind of busted. The tests do check for reset on sites changed in conjunction with an API key change, which is likely a wholesale change of setup, and that's the one where API usage should get reset. The benign change use case is, well, benign, and checks out in the real world now.

If I were truly AR I would build a specific test, and I may one day. It's on the list.

That or it'll be committed soon. Who can say? It may just be that I am completely AR and can't let it go...

autoSteve commented:
I may, or may not be AR... 😅

Test added.

I need to put this down and think of something new to do. Gonna send it.

"Something new" is most likely going to be detection of Amber and similar users avoiding any export when wholesale goes negative, which would probably throw off auto-dampen in times of regular Summer sun-shinyness. That's kind-of like an export limit scenario, but in that case PV production will equal site usage with zero site export.

The simple answer to that is to add a site usage entity into the mix. The harder answer is, "Are these entity values actually going to match, or is there going to be a slight real-world measurement difference between Enphase and Tesla? There is for me... yet the options dialogue grows.

I do so wish that HA would add the ability to simply split out an "advanced" config thingamy, and not just a step from one dialogue in the flow to another.
